include/ensmallen_bits/ada_bound/ams_bound_update.hpp

```diff
@@ -0,0 +1,214 @@
+  double& Epsilon() { return epsilon; }
+
+  //! Get the smoothing parameter.
+  double Beta1() const { return beta1; }
+  //! Modify the smoothing parameter.
+  double& Beta1() { return beta1; }
+
+  //! Get the second moment coefficient.
+  double Beta2() const { return beta2; }
+  //! Modify the second moment coefficient.
+  double& Beta2() { return beta2; }
+
+  //! Get the current iteration number.
+  size_t Iteration() const { return iteration; }
+  //! Modify the current iteration number.
+  size_t& Iteration() { return iteration; }
+
+  /**
+   * The UpdatePolicyType policy classes must contain an internal 'Policy'
+   * template class with two template arguments: MatType and GradType.  This is
+   * instantiated at the start of the optimization, and holds parameters
+   * specific to an individual optimization.
+   */
+  template<typename MatType, typename GradType>
+  class Policy
+  {
+   public:
+    /**
+     * This constructor is called by the SGD Optimize() method before the start
+     * of the iteration update process.
+     *
+     * @param parent AMSBoundUpdate object.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(AMSBoundUpdate& parent, const size_t rows, const size_t cols) :
+        parent(parent), first(true), initialStepSize(0)
+    {
+      m.zeros(rows, cols);
+      v.zeros(rows, cols);
+      vImproved.zeros(rows, cols);
+    }
+
+    /**
+     * Update step for AMSBound.
+     *
+     * @param iterate Parameters that minimize the function.
+     * @param stepSize Step size to be used for the given iteration.
+     * @param gradient The gradient matrix.
+     */
+    void Update(MatType& iterate,
+                const double stepSize,
+                const GradType& gradient)
+    {
+      // Convenience typedefs.
+      typedef typename MatType::elem_type ElemType;
+
+      // Save the initial step size.
+      if (first)
+      {
+        first = false;
+        initialStepSize = stepSize;
+      }
+
+      // Increment the iteration counter variable.
+      ++parent.iteration;
+
+      // Decay the first and second moment running average coefficient.
+      m *= parent.beta1;
+      m += (1 - parent.beta1) * gradient;
+
+      v *= parent.beta2;
+      v += (1 - parent.beta2) * (gradient % gradient);
+
+      const ElemType biasCorrection1 = 1.0 - std::pow(parent.beta1,
+          parent.iteration);
+      const ElemType biasCorrection2 = 1.0 - std::pow(parent.beta2,
+          parent.iteration);
+
+      const ElemType fl = parent.finalLr * stepSize / initialStepSize;
+      const ElemType lower = fl * (1.0 - 1.0 / (parent.gamma *
+          parent.iteration + 1));
+      const ElemType upper = fl * (1.0 + 1.0 / (parent.gamma *
+          parent.iteration));
+
+      // Element wise maximum of past and present squared gradients.
+      vImproved = arma::max(vImproved, v.diag());
```

Comment:
I think everything is correct except this, so if you want to revert 91b7653 or point out what I missed, then I think this is ready for merge. :+1: